File: bnovate/bnovate/report/subscription_invoices/subscription_invoices.py
# Copyright (c) 2023, bNovate, libracore and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _

def execute(filters=None):
    return get_columns(), get_data(filters)

def get_columns():
    return [
        {'fieldname': 'customer', 'label': _('Customer'), 'fieldtype': 'Link', 'options': 'Customer', 'width': 100},
        {'fieldname': 'customer_name', 'label': _('Customer name'), 'fieldtype': 'Data', 'width': 150},
        {'fieldname': 'subscription', 'label': _('Subscription'), 'fieldtype': 'Link', 'options': 'Subscription Contract', 'width': 100},
        {'fieldname': 'posting_date', 'label': _('Posting Date'), 'fieldtype': 'Date', 'width': 100},
        {'fieldname': 'sales_invoice', 'label': _('Invoice'), 'fieldtype': 'Link', 'options': 'Sales Invoice', 'width': 100},
        # {'fieldname': 'docstatus', 'label': _('Status'), 'fieldtype': 'Data', 'width': 100},
        {'fieldname': 'status', 'label': _('Status'), 'fieldtype': 'Data', 'width': 100},
    ]


def get_data(filters):
    subscription = filters.subscription if filters.subscription else "%"
    customer = filters.customer if filters.customer else "%" 
    sql_query = """
    SELECT
        sii.subscription,
        -- ss.name AS subscription,
        ss.customer,
        c.customer_name AS customer_name,
        si.posting_date AS posting_date,
        si.name AS sales_invoice,
        si.docstatus AS docstatus,
        si.status AS status
    FROM `tabSales Invoice Item` sii
        JOIN `tabSales Invoice` si ON sii.parent = si.name
        JOIN `tabSubscription Contract` ss ON ss.name = sii.subscription
        JOIN `tabCustomer` c on ss.customer = c.name
    WHERE ss.name LIKE "{subscription}"
        AND ss.customer LIKE "{customer}"
    ORDER BY ss.customer, ss.name, si.posting_date
    """.format(subscription = subscription, customer=customer)

    entries = frappe.db.sql(sql_query, as_dict=True)
    return entries

File: bnovate/bnovate/report/subscription_invoices/subscription_invoices.js
// Copyright (c) 2023, bNovate, libracore and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Subscription Invoices"] = {
    filters: [
        {
            "fieldname": "customer",
            "label": __("Customer"),
            "fieldtype": "Link",
            "options": "Customer"
        },
        {
            "fieldname": "subscription",
            "label": __("Subscription"),
            "fieldtype": "Link",
            "options": "Subscription Contract"
        },
    ],
    formatter(value, row, col, data, default_formatter) {
        // Copied from SINV's listview get_indicator
        var status_color = {
            "Draft": "grey",
            "Unpaid": "orange",
            "Paid": "green",
            "Return": "darkgrey",
            "Credit Note Issued": "darkgrey",
            "Unpaid and Discounted": "orange",
            "Overdue and Discounted": "red",
            "Overdue": "red",
            "Cancelled": "red"
        };
        if (col.fieldname === "status") {
            let color = status_color[value] ? status_color[value] : "grey";
            return `<span class="indicator ${color}">${value}</span>`
        }
        return default_formatter(value, row, col, data);
    }

};
